plugins/outputs/mqtt/mqtt.go

```diff
@@ -91,24 +108,15 @@ func (m *MQTT) Write(metrics []telegraf.Metric) error {
 	if len(metrics) == 0 {
 		return nil
 	}
+
+	metricsmap := make(map[string][]telegraf.Metric)
 	hostname, ok := metrics[0].Tags()["host"]
 	if !ok {
 		hostname = ""
 	}
-
-	metricsmap := make(map[string][]telegraf.Metric)
-
+	tt := NewTopicNameGenerator(hostname, m.TopicPrefix, m.template)
```

Comment:
> How about defining one template-topic containing all the static fields like hostname, prefix, the template itself etc. This can be done in Init() once

I am not sure we really want to add a new parameter to the struct, when we can achieve the same objective of instantiation only once, but with no need to have a struct variable. Should I make it a struct variable @srebhan ?

Reply:
Can you please elaborate why you cannot create a new topic-generator in `Init()` with 
```go
    m.generator = NewTopicNameGenerator(m.TopicPrefix, m.Topic)
```
saving this instance, compiling the template _once_ and then just pass in the parameters in `m.generator.Generate(hostname, metric)`? This will avoid memory trashing of the template and is a nice segregation of concerns as the plugin code doesn't need to care on what the generator does.

Reply:
So, I am not yet a good Go developer, memory allocation isn't my strength here, and I think I didn't get your point the first time. I believe I know get this better. By moving this to the `init()` we would be do memory allocation only once for the generator, because the `Write()` method is invoked after each metric is received, however, the `init()` only runs when we are bootstrapping telegraf.

Will follow advice 👍